### PR TITLE
Azure Stack IPI Support

### PR DESCRIPTION
**Pull request by patrickdillon**
Adds Terraform configuration to build Azure Stack clusters. Successfully creates all infrastructure, but cluster bootstrapping depends on further upstream changes.

Currently marked as WIP because:

- [x] Need to revendor for OpenShift fork of Terraform Provider. That can happen now that this has been tested succesfully
- [x] Align Terraform configuration with new staged approach. We can begin this work now, but we may want to push that into a separate PR to allow parallel work.

This should unblock CI work in openshift/release#18796

### CODE REVIEW

**Comment by patrickdillon** [comment]
cc @rna-afk

**Comment by patrickdillon** [comment]
I have removed WIP from this. I would like to merge this soon to enable parallel work and we'll bring the multistages into a separate PR.

**Comment by patrickdillon** [comment]
Depends on #5032
Dependency is loose but #5032 looks very close to merge. 
@rna-afk this was a pretty big rebase. hopefully it doesn't make your downstream changes too complicated. ash TF configs are in 8ccaee584bcbb99f7ef72444fc03705123899ff1

**Comment by patrickdillon** [comment]
I need to run another test to make sure bootstrap destroy works (which it did before switching to stages).

This is ready for review but I'm going to put a /hold until we get #5032 in or at least for today. 
/hold

File: data/data/azurestack/bootstrap/main.tf [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  client_id       = var.azure_client_id
+  client_secret   = var.azure_client_secret
+  tenant_id       = var.azure_tenant_id
+}
+
+resource "azurestack_storage_container" "ignition" {
+  name                  = "ignition"
+  resource_group_name   = var.resource_group_name
+  storage_account_name  = var.storage_account.name
+  container_access_type = "blob"
```
Comment (rna-afk) [inline data/data/azurestack/bootstrap/main.tf:24]: Matthew had a comment that suggested setting `container_access_type` to "private" if the `azurestack_storage_account_sas` resource was available in terraform provider. I added this to the provider so maybe we need to add that too?

File: data/data/azurestack/vnet/nsg.tf [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "*"
+  destination_port_range      = "6443"
+  source_address_prefix       = "*"
+  destination_address_prefix  = "*"
+  resource_group_name         = data.azurestack_resource_group.main.name
+  network_security_group_name = azurestack_network_security_group.cluster.name
+  description                 = local.description
+}
```
Comment (rna-afk) [inline data/data/azurestack/vnet/nsg.tf:19]: Matthew's work also included the `azurestack_subnet_network_security_group_association` which I added to the terraform provider. I could add all of these later too if you like.

**Review by rna-afk — COMMENTED** [review]
I think Matthew had 3 resources in his IPI work which had to be added to the terraform provider. The `azurestack_image` is added but the 2 others are missing. I can add those to the IPI work later or we could add them here if you like.

**Comment by patrickdillon** [comment]
/hold cancel
Azure PR is in merge pool.

**Comment by patrickdillon** [comment]
/hold

still waiting on #5032

**Comment by patrickdillon** [comment]
This is ready for review merge. I added two new commits for DNS labels and making bootstrap ignition private.

I was unable to get the CNAME records to resolve so we can follow up and switch DNS records in a subsequent commit.

**Comment by jhixson74** [comment]
This looks good to me. My only concern is changing the Azure profiles to older versions. Do we know the ramifications of this? I haven't yet looked, but I am certain we need some more recent versions for features such as private DNS. It would be nice to remove our forked version of that at some point.

**Comment by patrickdillon** [comment]
> This looks good to me. My only concern is changing the Azure profiles to older versions. Do we know the ramifications of this? I haven't yet looked, but I am certain we need some more recent versions for features such as private DNS. It would be nice to remove our forked version of that at some point

Right. I wasn't able to switch the private DNS API to the hybrid profile because private DNS is not offered in Azure Stack. I believe that was the only case where I was not able to make the change. 

None of the code was changed, just the API imports, so due to backward compatibility, this should not cause any regressions on Azure, while ensuring compatibility with Azure Stack.

**Comment by patrickdillon** [comment]
e2e-aws-upgade passed here: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/5084/pull-ci-openshift-installer-master-e2e-aws-upgrade/1418576833657442304

But the bot shows it is stuck trying to retest:
Screenshot from 2021-07-23 14-58-37
/override e2e-aws-upgrade

**Comment by patrickdillon** [comment]
Still in the same state. I'm probably just being impatient.

**Comment by patrickdillon** [comment]
/lgtm
Rebased conflict

**Pull request merged**